### PR TITLE
Fix pull not symlinking new and renamed files

### PR DESCRIPTION
**Pull request by yut23**
This PR reworks the way `pull` checks for new files for symlinking, which was broken before due to a missing `$` in front of the arithmetic expansion. This made it resolve to HEAD instead of HEAD@{1.seconds.ago}.
It now creates a temporary tag in each castle before running git pull, and diffs against that in symlink_new_files. It also wasn't checking for renamed files, but that's easily fixed with `--diff-filter=AR`.

This should be much more robust than the previous time-based method, which broke intermittently in my local tests and once during actual use. If `$now == $T_START` and the current second increments between `$now` and the git diff call, then HEAD@{1.seconds.ago} will point to the pull itself rather than the commit before it.

I've also added a whole bunch of tests covering all the error conditions and edge cases I could think of.

### CODE REVIEW

**Comment by andsens** [comment]
Welcome @yut23. Now those are some high quality PRs right there. Thank you very much.  
Now I'm wondering why I never just saved the current HEAD commit sha and compared against that though. I'm guessing it was too much bookkeeping when you are pulling multiple castles🤔 ...

The one issue I see is that something goes wrong between the tag creation and deletion, resulting in the check failing later on because of an internal rather than external conflict. No matter how you look at it though, this actually makes things work better than before. So I'm inclined to just merge it unless you want to address that issue first?

**Comment by yut23** [comment]
I originally wrote this code 2 years ago (!), and have been using it in my personal fork since then. I was thinking about that potential issue too, but I've never run into it, so it's probably not a huge problem.

**Comment by yut23** [comment]
OK, I've figured out how the tag gets cleaned up. `pull` is run inside a subshell in bin/homeshick, so `err` will just exit from that subshell. As long as the exit code isn't EX_USAGE, `symlink_cloned_files` will always be run, and the first thing it does is parse the tag and delete it.

I think this could fail if the user pulls two repos, the second of which already has the tag. This would make pull return EX_USAGE, and the tag in the first repo wouldn't be deleted. I'll see if I can write a test for that. Returning EX_DATAERR instead of EX_USAGE should fix it, and probably makes a bit more sense.

**Comment by yut23** [comment]
Hmm, that doesn't work. It looks like I was returning EX_USAGE specifically so symlink_cloned_files wouldn't delete a user-created tag.

File: lib/commands/pull.sh [hunk 1]
```diff
@@ -13,6 +14,15 @@ pull() {
     return "$EX_SUCCESS"
   fi
 
+  # this tag is exceedingly unlikely to already exist, but if it does, error
+  # out and let the user resolve it
+  (cd "$repo" && git rev-parse --verify "refs/tags/$BEFORE_PULL_TAG" &>/dev/null) && \
+    err "$EX_DATAERR" "Pull marker tag ($BEFORE_PULL_TAG) already exists in $repo. Please resolve this before pulling."
```
Comment (andsens) [inline lib/commands/pull.sh:20]: Actually, couldn't this be just a warning instead of killing everything? Especially when pulling multiple castles there's no reason to ruin the fun for everybody else :-)

File: lib/commands/pull.sh [hunk 2]
```diff
@@ -35,13 +46,15 @@ symlink_new_files() {
...
+      continue
+    fi
     if [[ ! -d $repo/home ]]; then
       continue;
     fi
     local git_out
-    local now
-    now=$(date +%s)
-    if ! git_out=$(cd "$repo" && git diff --name-only --diff-filter=A "HEAD@{(($now-$T_START+1)).seconds.ago}" HEAD -- home 2>/dev/null | wc -l 2>&1); then
+    if ! git_out=$(cd "$repo" && git diff --name-only --diff-filter=AR "$before_pull" HEAD -- home 2>/dev/null | wc -l 2>&1); then
```
Comment (andsens) [inline lib/commands/pull.sh:57]: Why not just diff to the tag and then delete it after? It's one less command that can go wrong.

File: lib/commands/pull.sh [hunk 3]
```diff
@@ -13,6 +14,15 @@ pull() {
...
   fi
 
+  # this tag is exceedingly unlikely to already exist, but if it does, error
+  # out and let the user resolve it
+  (cd "$repo" && git rev-parse --verify "refs/tags/$BEFORE_PULL_TAG" &>/dev/null) && \
+    err "$EX_DATAERR" "Pull marker tag ($BEFORE_PULL_TAG) already exists in $repo. Please resolve this before pulling."
+  # make a tag at the current commit, so we can compare against it below
+  (cd "$repo" && git tag "$BEFORE_PULL_TAG" 2>&1)
+  # remove the tag if one of the git operations fails
+  trap '{ cd "$repo" && git tag -d "$BEFORE_PULL_TAG" &>/dev/null; }' EXIT
```
Comment (andsens) [inline lib/commands/pull.sh:24]: Why the curly brackets?

File: lib/commands/pull.sh (hunk 1 above)
Comment (yut23) [reply lib/commands/pull.sh:20]: That should be how it works now, since pull is in a subshell. I'll add another test later.

File: lib/commands/pull.sh (hunk 2 above)
Comment (yut23) [reply lib/commands/pull.sh:57]: I think it was to make sure the tag still gets deleted in case `$repo/home` isn't a directory, but it looks like we can just move that check after the diff.

**Comment by yut23** [comment]
I've also got an alternate implementation in https://github.com/yut23/homeshick/tree/fix-pull-symlinking-old, which checks the git hash before and after calling pull in `bin/homeshick` and only calls symlink_new_files on the repos that changed. I thought the tag method was more elegant initially, but it does introduce more edge cases and error handling.

**Comment by andsens** [comment]
*ping. Do you think this is ready or would you like some more time to work on the PR?

**Comment by yut23** [comment]
I think it's ready. I've been using it for a while now with no issues.
